File: StacksAndQueue/206-reverse-linked-list/reverse-linked-list.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def reverseList(self, head):
        """
        :type head: Optional[ListNode]
        :rtype: Optional[ListNode]
        """
        if head == None or head.next== None:
            return head
        return self._reverse(head,None)
        
    def _reverse(self,head,prev):
        if(head == None):
            return prev
        front = head.next
        head.next = prev
        return self._reverse(front,head)

File: StacksAndQueue/206-reverse-linked-list/reverse-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseList(ListNode head) {
        if (head == null || head.next == null) return head;
        head = rev(head,null);
        return head;

    }
    public ListNode rev(ListNode head,ListNode prev){
        if(head == null) return prev;
        ListNode front = head.next;
        head.next = prev;
        head = rev(front,head);
        return head;
    }
}